gitpod-io/gitpod
# [installer]: deprecate static message bus password and replace with secret

## Description

The experimental config had a `staticMessagebusPassword` key which violates our rules on configuration, namely "[sensitive data must be stored in a secret](https://github.com/gitpod-io/gitpod/tree/main/install/installer/pkg/config#sensitive-data-must-be-stored-as-a-secret-not-in-plain-text)".

To that end, this deprecates the old behaviour and replaces with a new `messageBus` object in the config. I took the decision to put it in a new object because, as a dependency, it's plausible that we'll need to have additional configuration options in the future (similar to registry/db/storage).

The complexity with this ticket is that RabbitMQ load definitions secret which requires the password. As there's no documentation on how to achieve that, the load definition secret stores the password as `%PASSWORD%`, I've created an init container which reads the load definition and password secrets and replaces `%PASSWORD%` with the actual password and then saves it to the `emptyDir` which is what's actually used by the RabbitMQ resource.

If no password is given or the old `experimental` config is used, this stores the given password in a secret so RabbitMQ can read it.

There is no `MapValue` for the experimental config that can be done because it's too complex to create a secret for a time-limited deprecated function. The deprecated message doesn't display the password because I don't want to write credentials into logs, even if the password is easily readable.

## Related Issue(s)

Fixes #15899

## How to test

Install a cluster with the following config. To check that the new secret has been promulgated, run `kubectl exec -it -n gitpod deployments/server -- env | grep MESSAGEBUS_PASSWORD`

### 1. No password specified (default)

No special config required

### 2. Deprecated `experimental` config

```yaml
experimental:
  common:
    staticMessagebusPassword: some-password
```

### 3. New password secret

```shell
kubectl create secret generic -n gitpod --from-literal=rabbitmq-password=secret-password message-bus-password
```

```yaml
messageBus:
  credentials:
    kind: secret
    name: message-bus-password
```

## Release Notes

```release-note
[installer]: deprecate static message bus password and replace with secret 
```

## Documentation


## Build Options:

      Experimental feature to run the build with GitHub Actions (and not in Werft).
      If enabled this will build `install/preview`
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
